R/ebird_conn.R

```diff
@@ -46,35 +49,38 @@ ebird_conn <- function(dataset = c("observations", "checklists"),
   
   # check for a cached connection
   conn <- mget("birddb", envir = birddb_cache, ifnotfound = NA)[["birddb"]]
-  if (inherits(conn, "DBIConnection")) {
-    if (DBI::dbIsValid(conn)) {
-      # if view not already in databse, create it
-      if (!dataset %in% DBI::dbListTables(conn)) {
-        DBI::dbSendQuery(conn, view_query)
-      }
-      return(conn)
-    } else {
-      # shut down invalid cached connection to allow re-connection
-      DBI::dbDisconnect(conn, shutdown = TRUE)
-    }
-  }
   
-  # TODO: consider persisting connection on disk to avoid recreating view
-  conn <- DBI::dbConnect(drv = duckdb::duckdb(), ebird_db_dir())
+  # Disconnect if it's an invalid connection (expired in cache)
+  if ( db_is_invalid(conn) )  
+    conn <- DBI::dbDisconnect(conn, shutdown = TRUE)
+  
+  # We don't have a valid cached connection, so we must create one! 
+  if (!inherits(conn, "DBIConnection")){
+    conn <- DBI::dbConnect(drv = duckdb::duckdb(), ebird_db_dir())  
+  }
   
-  # set memory limit
+  # enforce memory limit
   # TODO: temp approach for testing, improve approach in production
   DBI::dbExecute(conn = conn, 
                  paste0("PRAGMA memory_limit='", memory_limit, "GB'"))
+
+  # create the view if does not exist
+  if (!dataset %in% DBI::dbListTables(conn)){
+    DBI::dbSendQuery(conn, view_query)
```

Comment:
I defined `view_query` at the top of the function since in my more convoluted logic I had two calls to `DBI::dbSendQuery(conn, view_query)`, but seems like now it should be brought down here since you only have the one use of it.

Reply:
good point! makes sense to keep the query code near where it is run.